src/flint/types/fmpz_mpoly.pyx

```diff
@@ -242,244 +404,399 @@ cdef class fmpz_mpoly(flint_mpoly):
                 fmpz_mpoly_divrem(res.val, res2.val, (<fmpz_mpoly>self).val, (<fmpz_mpoly>other).val, res.ctx.val)
                 return (res, res2)
         return NotImplemented
 
+    def __rdivmod__(self, other):
+        cdef fmpz_mpoly res, res2
+        if not self:
+            raise ZeroDivisionError("fmpz_mpoly divison by zero")
+        other = any_as_fmpz(other)
+        if other is not NotImplemented:
+            other = fmpz_mpoly(other, self.ctx)
+            res = create_fmpz_mpoly(self.ctx)
+            res2 = create_fmpz_mpoly(self.ctx)
+            fmpz_mpoly_divrem(res.val, res2.val, (<fmpz_mpoly>other).val, (<fmpz_mpoly>self).val, res.ctx.val)
+            return res
+        return NotImplemented
+
     def __floordiv__(self, other):
         cdef fmpz_mpoly res
-        if typecheck(self, fmpz_mpoly):
-            if typecheck(other, fmpz_mpoly):
-                if (<fmpz_mpoly>self).ctx is not (<fmpz_mpoly>other).ctx:
-                    ctx, (self, other) = coerce_fmpz_mpolys(self, other)
-                res = fmpz_mpoly.__new__(fmpz_mpoly)
-                res.ctx = (<fmpz_mpoly>self).ctx
-                fmpz_mpoly_init(res.val, res.ctx.val)
-                res._init = True
+        if typecheck(other, fmpz_mpoly):
+            if not other:
+                raise ZeroDivisionError("fmpz_mpoly division by zero")
+            if (<fmpz_mpoly>self).ctx is not (<fmpz_mpoly>other).ctx:
+                return NotImplemented
+            res = create_fmpz_mpoly(self.ctx)
+            fmpz_mpoly_div(res.val, (<fmpz_mpoly>self).val, (<fmpz_mpoly>other).val, res.ctx.val)
+            return res
+        else:
+            other = any_as_fmpz(other)
+            if other is not NotImplemented:
+                if not other:
+                    raise ZeroDivisionError("fmpz_mpoly division by zero")
+                other = fmpz_mpoly(other, self.ctx)
+                res = create_fmpz_mpoly(self.ctx)
                 fmpz_mpoly_div(res.val, (<fmpz_mpoly>self).val, (<fmpz_mpoly>other).val, res.ctx.val)
                 return res
         return NotImplemented
 
+    def __rfloordiv__(self,other):
+        cdef fmpz_mpoly res
+        if not self:
+            raise ZeroDivisionError("fmpz_mpoly division by zero")
+        other = any_as_fmpz(other)
+        if other is not NotImplemented:
+            other = fmpz_mpoly(other, self.ctx)
+            res = create_fmpz_mpoly(self.ctx)
+            fmpz_mpoly_div(res.val, (<fmpz_mpoly>other).val,  self.val, res.ctx.val)
+            return res
+        return NotImplemented
+
+    def __truediv__(self, other):
+        cdef fmpq_mpoly qres
+        cdef fmpq_t q_other
+        cdef int xtype
+
+        if typecheck(other, fmpz_mpoly):
+            if not other:
+                raise ZeroDivisionError("fmpz_mpoly division by zero")
+            if self.ctx is not (<fmpz_mpoly> other).ctx:
+                return NotImplemented
+            return fmpz_mpoly_q(self, other)
```

Comment:
We need a more general decision on how to handle `a / b` in non-field rings.

Currently this is a TypeError in all other cases:
```python
In [2]: fmpz(2) / fmpz(2)
...
TypeError: unsupported operand type(s) for /: 'flint.types.fmpz.fmpz' and 'flint.types.fmpz.fmpz'
```
Likewise all polynomials and matrices. Division by a scalar is allowed where the base ring is a field e.g.:
```python
In [4]: fmpz_poly([1, 0]) / fmpz(1)
...
TypeError: unsupported operand type(s) for /: 'flint.types.fmpz_poly.fmpz_poly' and 'flint.types.fmpz.fmpz'

In [5]: fmpq_poly([1, 0]) / fmpq(1)
Out[5]: 1
```
There is one exception:
```python
In [8]: fmpz_mat([[1], [0]]) / fmpz(2)
Out[8]: 
[1/2]
[  0]
```
In other words scalar division of an `fmpz_mat` by an `fmpz` returns an `fmpq_mat`.

There is one other case of implicit conversion to a larger domain:
```python
In [15]: fmpz_mat([[2, 0], [0, 2]]).inv()
Out[15]: 
[1/2,   0]
[  0, 1/2]
```

Reply:
Ok, I think you're probably right.  The gains in convenience are probably not worth the loss of sanity.

Reply:
I'm not sure what is best here in general. At least in the places where SymPy would use this the code is very careful not to use `a / b` if it is not known that we are working in a field. It would look like:
```python
In [1]: from sympy import *

In [2]: x, y = symbols('x, y')

In [3]: R = QQ[x,y]

In [4]: R
Out[4]: QQ[x,y]

In [5]: p1 = R(x - y)

In [6]: p2 = R(x + y)

In [7]: K = R.get_field()

In [8]: K
Out[8]: QQ(x,y)

In [9]: K.convert_from(p1, R) / K.convert_from(p2, R)
Out[9]: (x - y)/(x + y)
```
Currently lacking in python-flint to make this work is:

- A systematic way to get the field of fractions from a ring.
- A systematic way to convert from the ring to that field.
- A systematic way to get the numerator and denominator back from the field as elements of the ring of integers.

It is not an immediate problem for SymPy that these things are lacking but I think that for someone using python-flint directly there should be some way to do this.

Probably what is needed is to have ring objects for every domain (including those that don't have a context in Flint like fmpz, fmpq, nmod). Probably this is all something that would be easier with generic rings so I am unsure right now what is the best approach to take in python-flint for conversions.

It might be good to have a meeting at some point to talk about some of these basic design questions.

Reply:
IMO the ``/`` operator should work as ``div`` in the new generics system: preserve the type, allow divisions when the quotient exists in the ring, raise an exception otherwise.

Reply:
> preserve the type, allow divisions when the quotient exists in the ring, raise an exception otherwise.

That seems reasonable to me. Also if it is something that is used consistently in generic rings then it would be good to match that here and to use it consistently also across python-flint.

In SymPy this kind of division is called `exquo`:
```python
In [10]: R.exquo(p1, p2)
...
ExactQuotientFailed: x + y does not divide x - y in QQ[x,y]

In [11]: R.exquo(p1, p1)
Out[11]: 1
```
This is a useful operation that is used in e.g. fraction-free algorithms. Exposing it as `a / b` would give SymPy a direct way to access the operation from python-flint rather than needing to use `divmod`.

It is unfortunate that SymPy's domain types cannot be trusted to use `a / b` for exquo but it all starts from using `int` or `gmpy2.mpz` for `ZZ`. Having this for `fmpz` leaves it still a little incompatible with `int` and `gmpy2.mpz` but that is unavoidable unless we want to match what they do:
```python
In [12]: 2/3
Out[12]: 0.6666666666666666
```
I think it is fine for `fmpz` to be incompatible with this because `gmpy2.mpz` already exists for someone who wants a dropin faster replacement for `int`.

Reply:
A meeting would be great.  This week is not great as I have people fixing my roof, but after that anytime after 1200 GMT would do.

I also have type related questions in reference to the __call__ method of the various mpoly types:  should I insist that all the parameters have identical types, or that at  least one of the parameters is of a type that all the others can be promoted to (for example if one was an arb_poly then the other parameters could be arb, fmpz or fmpz_poly), or that there was some type, not necessarily occuring in the list, that all the parameters could be converted to.

I do think that we should probably find a way of inserting the flint generics under python-flint.  Issue #86 addresses this as well.

Reply:
Sure.

Reply:
Yeah that sounds good. Happy to make time next week. I'm in the UK re: timezones and planning but happy to be flexible :)

Reply:
Great, I'll follow up by email.

I will just note here some of the questions that I would like to resolve or at least make progress on:

- How exactly python-flint should use Flint's generics and also how to handle things that are not really generic.
- What a higher level python-flint interface might look like.
- How to handle explicit conversions between different types/domains.
- What to do with operations that do not necessarily make sense in a given domain like division and inverse.